extensions/ql-vscode/src/language-support/contextual/template-provider.ts

```diff
@@ -284,15 +286,17 @@ export class TemplatePrintCfgProvider {
   }
 
   async provideCfgUri(
-    document?: TextDocument,
+    document: TextDocument,
+    line: number,
+    character: number,
   ): Promise<[Uri, Record<string, string>] | undefined> {
-    if (!document) {
-      return;
-    }
-
     return this.shouldUseCache()
-      ? await this.cache.get(document.uri.toString())
-      : await this.getCfgUri(document.uri.toString());
+      ? await this.cache.get(
+          `${document.uri.toString()}#${line}:${character}`,
+          line,
+          character,
+        )
```

Comment:
So, now we will be caching by line/character/path. Will the cache blow up if someone requests the same CFG but on different caret locations on the same line?

Since this feature is mostly used for debugging CFG implementations, I wonder if it is even useful to use the cache at all? Presumably most users of this graph will run with caching disabled.

Reply:
I don't think the cache is very useful ; I just added the line an column in the cache-key to avoid returning the same CFG each time regardless of where the cursor is. The cache seems to have LRU strategy, so hopefully it should not blow up and evict old results. We can also disable caching together.

Reply:
I think it might still be useful to have the cache. Even though you could click in arbitrary places in a method, as a user of the feature I'd most likely right-click on the method name so the chance of hitting the cache if perhaps not that low.